content/extending-chat-widget/custom-identity-provider/index.mdx

```diff
@@ -0,0 +1,174 @@
+- `hasToken` - resolving boolean. Determining whether a token has been acquired.
+- `invalidate` - resolving nothing. When called should remove the current token. There is no need to do anything else, as a new token will be requested by `getFreshToken` afterward.
+
+# Chat Widget Token
+
+LiveChat authorization token you are willing to resolve in functions should contain the following properties:
+
+| Parameter      | Description                                                                      |
+| -------------- | -------------------------------------------------------------------------------- |
+| `accessToken`  | A token you can use to call LiveChat APIs on behalf of the user (customer).      |
+| `entityId`     | The ID of the customer.                                                              |
+| `expiresIn`    | A number in **milliseconds** specifying how long the `accessToken` will be valid.|
+| `tokenType`    | Value: `Bearer`.                                                                 |
+| `creationDate` | Unix timestamp specyfing the time of creation of the token.                       |
+| `licenseId`    | LiveChat license ID.                                                             |
+
+## How to acquire it?
+
+In order to get the token you need to implement authorizing function making a call to a `/customer` endpoint from LiveChat Accounts API. To do so, you can follow [creating a new customer flow](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#creating-a-new-customer) or have a look at our [example](https://github.com/livechat/identity-provider-example/blob/master/lib/get-customer-token.ts) using cookies. Mind the fact that the [data](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#response-4) returned from this endpoint is not 1:1 with the token shape described above. You will have to add the `creationDate` and `licenseId` fields. Additionally, we expect `expiresIn` to be defined in milliseconds while the endpoint operates in seconds. Therefore you should multiply the provided value by `1000` before passing it further.
+
+
+# Examples
+
+Here are our suggested implementations of Customer Identity Provider. The `fetchLiveChatToken()` method used in examples is used to acquire the [Chat Widget token](#chat-widget-token).
+
+## Without caching
+
+If you don't need to cache the token, the fetching logic shall occur on every refresh of your application page. In such a case, the implementation can be very simple - we only have to ensure handling of `getToken` promise and simply resolve others.
+
+```js
+ window.__lc.custom_identity_provider = () => {
+    const fetchLiveChatToken = () => {
+        // fetch a token from LiveChat Accounts API here
+    }
+
+    let tokenPromise = null
+    const fetchToken = () => {
+        tokenPromise = fetchLiveChatToken()
+            .then(response => {
```

Comment:
I feel that it's important to provide a comprehensive example here. We should use a catch handler here, clear the `tokenPromise` and rethrow the error (at the very least).

Reply:
Which also raises the need to document errors. We should either expect them to match the ones we create in the `@livechat/customer-auth` or we should learn to adapt whatever can be thrown here to our needs. I feel like probably the first one is better as it sets clear expectations, both on the consumer's and provider's sides of the story.

Reply:
good point I think we shouldn't overcomplicate it too much so rethrowing an error should be good enough 👍